src/Processors/Formats/Impl/Parquet/ParquetLeafColReader.cpp

```diff
@@ -488,17 +455,114 @@ void ParquetLeafColReader<TColumn>::readPageV1(const parquet::DataPageV1 & page)
+        auto bit_reader = std::make_unique<arrow::bit_util::BitReader>(buffer + 4, num_bytes);
+        num_bytes += 4;
+        buffer += num_bytes;
+        max_size -= num_bytes;
+        def_level_reader = std::make_unique<RleValuesReader>(std::move(bit_reader), bit_width);
+    }
+    else
+    {
+        def_level_reader = std::make_unique<RleValuesReader>(page.num_values());
+    }
+
+    initDataReader(page.encoding(), buffer, max_size, std::move(def_level_reader));
+}
+
+/*
+ * As far as I understand, the difference between page v1 and page v2 lies primarily on the below:
+ * 1. repetition and definition levels are not compressed;
+ * 2. size of repetition and definition levels is present in the header;
+ * 3. the encoding is always RLE
+ *
+ * Therefore, this method leverages the existing `parquet::LevelDecoder::SetDataV2` method to build the repetition level decoder.
+ * The data buffer is "offset-ed" by rl bytes length and then dl decoder is built using RLE decoder. Since dl bytes length was present in the header,
+ * there is no need to read it and apply an offset like in page v1.
+ * */
+template <typename TColumn>
+void ParquetLeafColReader<TColumn>::readPageV2(const parquet::DataPageV2 & page)
+{
+    static parquet::LevelDecoder repetition_level_decoder;
+
+    cur_page_values = page.num_values();
+
+    const auto * buffer =  page.data();
+
+    const int64_t total_levels_length =
+        static_cast<int64_t>(page.repetition_levels_byte_length()) +
+        page.definition_levels_byte_length();
+
+    if (total_levels_length > page.size())
+    {
+        throw Exception(
+            ErrorCodes::BAD_ARGUMENTS, "Data page too small for levels (corrupt header?)");
+    }
+
+    if (col_descriptor.max_repetition_level() > 0)
+    {
+        repetition_level_decoder.SetDataV2(
+            page.repetition_levels_byte_length(),
+            col_descriptor.max_repetition_level(),
+            cur_page_values,
+            buffer);
+    }
+
+    // ARROW-17453: Even if max_rep_level_ is 0, there may still be
+    // repetition level bytes written and/or reported in the header by
+    // some writers (e.g. Athena)
+    buffer += page.repetition_levels_byte_length();
+
+    assert(col_descriptor.max_definition_level() >= 0);
+    std::unique_ptr<RleValuesReader> def_level_reader;
+    if (col_descriptor.max_definition_level() > 0)
+    {
+        auto bit_width = arrow::bit_util::Log2(col_descriptor.max_definition_level() + 1);
+        auto num_bytes = page.definition_levels_byte_length();
+        auto bit_reader = std::make_unique<arrow::bit_util::BitReader>(buffer, num_bytes);
+        def_level_reader = std::make_unique<RleValuesReader>(std::move(bit_reader), bit_width);
+    }
+    else
+    {
+        def_level_reader = std::make_unique<RleValuesReader>(page.num_values());
+    }
+
+    initDataReader(page.encoding(), &buffer[total_levels_length], page.size(), std::move(def_level_reader));
```

Comment:
1. There's `buffer += page.repetition_levels_byte_length()` above, so adding `total_levels_length` here makes no sense. Do `buffer += page.definition_levels_byte_length()` before this and use `buffer` here?
2. `page.size() - total_levels_length`

(I guess this works in tests because repetition levels are empty because arrays are not supported yet, and I guess data reader doesn't check if the given length is too long?)